modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -317,19 +335,222 @@ inline Point3f HashTSDFVolumeCPU::getNormalVoxel(Point3f point) const
-    //std::cout << normal << std::endl;
+
     float nv = sqrt(normal[0] * normal[0] +
                      normal[1] * normal[1] +
                      normal[2] * normal[2]);
     return nv < 0.0001f ? nan3 : normal/nv;
 }
 
+#else
+
+inline Point3f HashTSDFVolumeCPU::getNormalVoxel(Point3f point) const
+{
+    cv::Point3f neighbourCoords[] = {
+                                Point3f(0, 0, 0),
+                                Point3f(0, 0, 1),
+                                Point3f(0, 1, 0),
+                                Point3f(0, 1, 1),
+                                Point3f(1, 0, 0),
+                                Point3f(1, 0, 1),
+                                Point3f(1, 1, 0),
+                                Point3f(1, 1, 1) };
+    for (int i = 0; i < 8; i++)
+        neighbourCoords[i] *= voxelSize;
+
+    Point3f pointVec(point);
+    Point3f normal = Point3f(0, 0, 0);
+
+    Point3f c0_pointPrev = pointVec, c0_pointNext = pointVec;
+    Point3f c1_pointPrev = pointVec, c1_pointNext = pointVec;
+    Point3f c2_pointPrev = pointVec, c2_pointNext = pointVec;
+
+    c0_pointPrev.x -= voxelSize; c0_pointNext.x += voxelSize;
+    c1_pointPrev.y -= voxelSize; c1_pointNext.y += voxelSize;
+    c2_pointPrev.z -= voxelSize; c2_pointNext.z += voxelSize;
+
+    float c0_n_vx[8], c1_n_vx[8], c2_n_vx[8];
+    float c0_p_vx[8], c1_p_vx[8], c2_p_vx[8];
+
+    cv::Vec3i c0_n_vui[8], c0_p_vui[8];
+
+    for (int i = 0; i < 8; i++)
+        c0_n_vui[i] = volumeToVolumeUnitIdx(neighbourCoords[i] + c0_pointNext);
+    if (vuiCheck(c0_n_vui))
+    {
+        VolumeUnitMap::const_iterator it = volumeUnits.find(c0_n_vui[0]);
+        for (int i = 0; i < 8; i++)
+            c0_n_vx[i] = tsdfToFloat(_at_3arg(neighbourCoords[i] + c0_pointNext, c0_n_vui[0], it).tsdf);
+    }
+    else {
+        for (int i = 0; i < 8; i++)
+            c0_n_vx[i] = tsdfToFloat(_at_2arg(neighbourCoords[i] + c0_pointNext, c0_n_vui[i]).tsdf);
+    }
+
+    for (int i = 0; i < 8; i++)
+        c0_p_vui[i] = volumeToVolumeUnitIdx(neighbourCoords[i] + c0_pointPrev);
+    if (vuiCheck(c0_p_vui))
+    {
+        VolumeUnitMap::const_iterator it = volumeUnits.find(c0_p_vui[0]);
+        for (int i = 0; i < 8; i++)
+            c0_p_vx[i] = tsdfToFloat(_at_3arg(neighbourCoords[i] + c0_pointPrev, c0_p_vui[0], it).tsdf);
+    } else {
+        for (int i = 0; i < 8; i++)
+            c0_p_vx[i] = tsdfToFloat(_at_2arg(neighbourCoords[i] + c0_pointPrev, c0_p_vui[i]).tsdf);
+    }
+    for (int i = 0; i < 8; i++)
+        if (i % 4 > 1)
+            c1_n_vx[i] = tsdfToFloat(_at_1arg(neighbourCoords[i] + c1_pointNext).tsdf);
+        else
+            c1_p_vx[i] = tsdfToFloat(_at_1arg(neighbourCoords[i] + c1_pointPrev).tsdf);
+
+    for (int i = 0; i < 8; i++)
+        if (i % 2 == 1)
+            c2_n_vx[i] = tsdfToFloat(_at_1arg(neighbourCoords[i] + c2_pointNext).tsdf);
+        else
+            c2_p_vx[i] = tsdfToFloat(_at_1arg(neighbourCoords[i] + c2_pointPrev).tsdf);
+
```

Comment:
I think the code can be simplified & boosted in performance sense.
1. We can precalculate coordinate offsets of all 32 points we need instead of these complex for loops
1. We should check if **all** points we need are in the same volume unit just by checking point coordinate remainder.
2. If they are not, then the following procedure may happen:
* We can make a small array which works like a dictionary from volume unit index to corresponding iterator. We base it on the following:
 * there can't be more than 8 adjacent volume units
 * their indices can be easily mapped to 0 to 7 using coordinate parity:
  - `(x % 2) + (y % 2)*2 + (z % 2)*4`
* For each point we:
 * get its volume unit index
 * check if it's in our dictionary
 * if it's not, we make a hash table query and write a result into a dictionary
 * if it's in, we use it with that `at` function which takes an iterator
* then we rearrange these values into vectors and pass them into `interpolate()` calls

Reply:
done